plenum/server/replica.py

```diff
@@ -1045,6 +1055,15 @@ def process_three_phase_msg(self, msg: ThreePhaseMsg, sender: str):
         """
         sender = self.generateName(sender, self.instId)
 
+        pp_key = ((msg.viewNo, msg.ppSeqNo) if
+                  isinstance(msg, PrePrepare) else None)
+
+        # the same PrePrepare might come here multiple times
+        if (pp_key and msg not in self.pre_prepare_tss[pp_key]):
```

Comment:
`self.pre_prepare_tss[pp_key]` raises a KeyError if `pp_key` is not in `self.pre_prepare_tss` yet, doesn't it?